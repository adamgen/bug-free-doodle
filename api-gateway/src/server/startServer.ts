import { ApolloServer } from "apollo-server-express";
import config from "config";
import cookieParser from "cookie-parser";
import cors from "cors";
import express from "express";
import morgan from "morgan";
import resolvers from "#root/graphql/resolvers";
import schema from "#root/graphql";

import formatGraphQLErrors from "./formatGraphQLErrors";
import injectSession from "./middleware/injectSession";

const PORT = <number>config.get("PORT");

const startServer = async () => {
  const apolloServer = new ApolloServer({
    context: (ctx) => ({
      ...ctx,
      userSession: ctx.res.locals.userSession,
    }),
    formatError: formatGraphQLErrors,
    resolvers,
    typeDefs: schema,
  });

  const app = express();
  // app.use(morgan("dev"));
  app.use(cookieParser());
  // app.use(express.urlencoded({ extended: true }));
  app.use(express.json());
  app.use(
    cors({
      credentials: true,
      origin: (origin, cb) => cb(null, true),
    })
  );

  // app.use(injectSession);

  await apolloServer.start();
  apolloServer.applyMiddleware({
    app,
    cors: { credentials: true, origin: "https://studio.apollographql.com" },
    path: "/graphql",
  });

  app.listen(PORT, "0.0.0.0", () => {
    console.info(`API gateway listening on ${PORT}`);
  });
};

export default startServer;
