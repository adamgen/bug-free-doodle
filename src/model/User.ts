import { DataTypes, Model, UUIDV4 } from "sequelize";
import sequelize from "../config/sequelize";

interface UserAttributes {
  id: string;
  email: string;
  password: string;
}

export class User extends Model<UserAttributes> {}

User.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: UUIDV4,
      allowNull: false,
      primaryKey: true,
    },
    email: {
      type: DataTypes.STRING,
      allowNull: false,
      unique: true,
    },
    password: {
      type: DataTypes.STRING,
      allowNull: false,
    },
  },
  {
    sequelize: sequelize,
    tableName: "users",
  }
);
