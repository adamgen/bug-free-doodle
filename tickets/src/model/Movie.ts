import { DataTypes, Model, UUIDV4 } from "sequelize";
import sequelize from "../config/sequelize";

interface MovieAttributes {
  id: string;
  title: string;
  imdbApiID: string;
  length: number;
}

export class Movie extends Model<MovieAttributes> {
  public id!: number;
  public title!: string;
  public imdbApiID!: string;
  public length!: number;

  public readonly createdAt!: Date;
  public readonly updatedAt!: Date;
}

Movie.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: UUIDV4,
      allowNull: false,
      primaryKey: true,
    },
    title: {
      type: DataTypes.STRING,
      allowNull: false,
      unique: true,
    },
    imdbApiID: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    length: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
  },
  {
    sequelize: sequelize,
    tableName: "movie",
  }
);
