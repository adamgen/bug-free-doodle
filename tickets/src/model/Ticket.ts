import { DataTypes, Model, UUIDV4 } from "sequelize";
import sequelize from "../config/sequelize";

interface TicketAttributes {
  id: string;
  movieId: string;
  availableTickets: number;
  ticketsLeft: number;
  dateAndTIme: Date;
  price: number;
}

export class Ticket extends Model<TicketAttributes> {
  public id!: number;
  public movieId!: string;
  public availableTickets!: number;
  public ticketsLeft!: number;
  public dateAndTIme!: Date;
  public price!: number;

  public readonly createdAt!: Date;
  public readonly updatedAt!: Date;
}

Ticket.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: UUIDV4,
      allowNull: false,
      primaryKey: true,
    },
    movieId: {
      type: DataTypes.STRING,
      allowNull: false,
    },
    availableTickets: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
    ticketsLeft: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
    dateAndTIme: {
      type: DataTypes.DATE,
      allowNull: false,
    },
    price: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
  },
  {
    sequelize: sequelize,
    tableName: "ticket",
  }
);
