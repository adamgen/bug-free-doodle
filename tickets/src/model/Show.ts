import { DataTypes, Model, UUIDV4 } from "sequelize";
import sequelize from "../config/sequelize";
import { Ticket } from "./Ticket";
import { v4 as uuidv4 } from "uuid";

interface ShowAttributes {
  id: string;
  dateAndTIme: Date;
  price: number;
}

interface TicketsInit {
  ticketAmount: number;
  imdbApiId: string;
}

export class Show extends Model<ShowAttributes> {
  public id!: number;
  public dateAndTIme!: Date;
  public price!: number;

  public readonly createdAt!: Date;
  public readonly updatedAt!: Date;

  public async initTickets(
    ticketsInit: TicketsInit,
    showId: string
  ): Promise<any> {
    const { imdbApiId, ticketAmount } = ticketsInit;

    for (let i = 1; i <= ticketAmount; i++) {
      const id = uuidv4();
      const seatId = i;
      const isTaken = false;
      await Ticket.create({ id, showId, imdbApiId, seatId, isTaken });
    }
  }
}

Show.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: UUIDV4,
      allowNull: false,
      primaryKey: true,
    },
    dateAndTIme: {
      type: DataTypes.DATE,
      allowNull: false,
    },

    price: {
      type: DataTypes.INTEGER,
      allowNull: false,
    },
  },
  {
    sequelize: sequelize,
    tableName: "show",
  }
);
