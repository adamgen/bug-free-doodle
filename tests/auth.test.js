// Testing Authentication API Routes
// ðŸ’¯ Test all the edge cases

import axios from "axios";
import * as generate from "./utils/generate";
import {
  getData,
  getDataAndCookie,
  handleRequestFailure,
  resolve,
} from "./utils/async";
import * as usersDB from "../src/model/User";
import startServer from "../src/server";
import cookieParser from "cookie-parser";

// import { User } from "../model/User";

let api, server;

beforeAll(async () => {
  server = await startServer();
  server.use(cookieParser);
  const baseURL = `http://localhost:${process.env.PORT}/`;
  api = axios.create({ baseURL });
  // api.interceptors.response.use(getDataAndCookie, handleRequestFailure);
});

// afterAll(() => server.close());

//TODO
// import { resetDb } from "utils/db-utils";
// beforeEach(() => resetDb());
describe("Auth Flow", () => {
  const { email, password } = generate.loginForm();
  test("register", async () => {
    console.log(email, password);
    // register
    const rData = await api.post("auth/register", { email, password });
    expect(rData.status).toBe(200);
  });
  test("login", async () => {
    // login
    const lData = await api.post("auth/login", { email, password });
    expect(lData.status).toBe(200);

    //TODO authenticated request
    // const mData = await api.get("auth/me", {
    //   headers: {
    //     Authorization: `Bearer ${lData.user.token}`,
    //   },
    // });
    // expect(mData.user).toEqual({
    //   token: expect.any(String),
    //   id: expect.any(String),
    //   username,
    // });
  });
});

test("email must be unique", async () => {
  const { email, password } = generate.loginForm();
  const rData = await api.post("auth/register", { email, password });

  const error = await api
    .post("auth/register", { email, password })
    .catch((e) => e.response);
  expect(error.status).toBe(400);
  expect(error.data).toMatchSnapshot(`{message:'email taken'}`);
  // console.log("er", error);
  // expect(error.status).toBe(400);

  //   const error = await api
  //     .post("auth/register", { email, password: "Nancy-is-#1" })
  //     .catch(resolve);
  //   expect(error).toMatchInlineSnapshot(
  //     `[Error: 400: {"message":"email taken"}]`
  //   );
});
//TODO
// test("get me unauthenticated returns error", async () => {
//   const error = await api.get("auth/me").catch(resolve);
//   expect(error).toMatchInlineSnapshot(
//     `[Error: 401: {"code":"credentials_required","message":"No authorization token was found"}]`
//   );
// });

test("email required to register", async () => {
  const error = await api
    .post("auth/register", { password: generate.password() })
    .catch((e) => e.response);
  expect(error.status).toBe(400);
  expect(error.data).toMatchSnapshot(`{message: "email can't be blank"}`);
});

test("password required to register", async () => {
  const error = await api
    .post("auth/register", { email: generate.email() })
    .catch((e) => e.response);
  expect(error.status).toBe(400);
  expect(error.data).toMatchSnapshot(`{message:"password can't be blank"}`);
});

// test("email required to login", async () => {
//   const error = await api
//     .post("auth/login", { password: generate.password() })
//     .catch(resolve);
//   expect(error).toMatchInlineSnapshot(
//     `[Error: 400: {"message":"email can't be blank"}]`
//   );
// });

// test("password required to login", async () => {
//   const error = await api
//     .post("auth/login", { email: generate.email() })
//     .catch(resolve);
//   expect(error).toMatchInlineSnapshot(
//     `[Error: 400: {"message":"password can't be blank"}]`
//   );
// });

// test("user must exist to login", async () => {
//   const error = await api
//     .post("auth/login", generate.loginForm({ email: "__will_never_exist__" }))
//     .catch(resolve);
//   expect(error).toMatchInlineSnapshot(
//     `[Error: 400: {"message":"email or password is invalid"}]`
//   );
// });
