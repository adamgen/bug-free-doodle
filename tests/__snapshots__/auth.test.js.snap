// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`email must be unique: {message:'email taken'} 1`] = `
Object {
  "message": "email taken",
}
`;

exports[`email required to register: {message: "email can't be blank"} 1`] = `
Object {
  "message": "email can't be blank",
}
`;

exports[`password required to register: {message:"password can't be blank"} 1`] = `
Object {
  "message": "password can't be blank",
}
`;
